### PR TITLE
REN-56: Image hosting on AWS

### PR DESCRIPTION
**Pull request by AsadA128**

### CODE REVIEW

File: Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/FileUploadController.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    @RequestMapping("/{filename:.+}")
+    @ResponseBody
+    public ResponseEntity<Resource> serveFile(@PathVariable String filename) {
+        Resource file = fileStorageService.loadAsResource(filename);
+        return ResponseEntity.ok()
+                .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\""+file.getFilename()+"\"")
+                .body(file);
+    }
+
+    @RequestMapping(value = "/upload", method = RequestMethod.POST)
```
Comment (beeduck) [inline Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/FileUploadController.java]: Just /files POST, no /upload

File: Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    ListingImageService listingImageService;
+
+    @RequestMapping(value = "/upload", method = RequestMethod.POST)
+    public String handleFileUpload(@RequestParam("listingId") int listingId,
+                                   @RequestParam("files") MultipartFile[] files) throws IOException {
+        return listingImageService.uploadImage(listingId,files);
+    }
+
+//    @RequestMapping(value = "/{uuid}", method = RequestMethod.GET)
+//    public ResponseEntity<Resource> serveFile(@PathVariable("uuid") String uuid) throws IOException {
```
Comment (beeduck) [inline Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java]: Remove

File: Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+/**
+ * Created by Asad on 11/29/2016.
+ */
+
+@RestController
+@RequestMapping("/listingImages")
```
Comment (beeduck) [inline Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java]: "/listing-images"

File: Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+public class ListingImageController {
+
+    @Autowired
+    ListingImageService listingImageService;
+
+    @RequestMapping(value = "/upload", method = RequestMethod.POST)
```
Comment (beeduck) [inline Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java]: remove "/upload"

File: Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+//    public ResponseEntity<Resource> serveFile(@PathVariable("uuid") String uuid) throws IOException {
+//        return listingImageService.getImageById(uuid);
+//    }
+
+    @RequestMapping(value = "/{uuid}", method = RequestMethod.GET)
+    public void serveFile(HttpServletResponse response, @PathVariable("uuid") String uuid) throws IOException {
+        listingImageService.getImageByIdS3(response,uuid);
+    }
+
+    @RequestMapping(value = "/byListingId/{listingId}", method = RequestMethod.GET, produces = "application/zip")
```
Comment (beeduck) [inline Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java]: remove by listing ID

File: Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    public byte[] serveFileByListing(HttpServletResponse response, @PathVariable("listingId") int listingId) throws IOException {
+        return listingImageService.getImageByListingId(response, listingId);
+    }
+
+    @RequestMapping(value = "/deleteByUUID/{uuid}", method = RequestMethod.DELETE)
```
Comment (beeduck) [inline Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java]: just "/{uuid}"

File: Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    public void deleteByUUID(@PathVariable("uuid") String uuid) {
+        listingImageService.deleteByImageUUID(uuid);
+    }
+
+    @RequestMapping(value = "/deleteByListing/{listingId}", method = RequestMethod.DELETE)
```
Comment (beeduck) [inline Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java:47]: remove delete by listing ID

File: Rent/Rent-API/src/main/java/com/rent/api/services/listing/ListingImageServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        byte[] data = new byte[16384];
+        while ((nRead = stream.read(data, 0, data.length)) != -1) {
+            os.write(data, 0, nRead);
+        }
+        os.flush();
+        stream.close();
+        os.close();
+    }
+
+    public byte[] getImageByListingId(HttpServletResponse response, int listingId) throws IOException {
```
Comment (beeduck) [inline Rent/Rent-API/src/main/java/com/rent/api/services/listing/ListingImageServiceImpl.java]: Remove this if the controller for it is being removed?

File: Rent/Rent-API/src/main/java/com/rent/api/services/listing/ListingImageServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+    @Transactional
+    public void deleteByImageUUID(String uuid) {
+        String path = listingImageRepository.findByImageUUID(uuid).getPath();
+        AWSConnector.deleteImageFromS3(path);
+        listingImageRepository.deleteByImageUUID(uuid);
+    }
+
+    @Transactional
+    public void deleteByListingId(int listingId) {
```
Comment (beeduck) [inline Rent/Rent-API/src/main/java/com/rent/api/services/listing/ListingImageServiceImpl.java:126]: Remove if controller is removed?

File: Rent/Rent-API/src/main/resources/application.properties [hunk 10]
```diff
@@ -1,13 +1,20 @@
 # JACKSON
 spring.jackson.date-format = yyyy-MM-dd HH:mm:ss
 
-# LOGGING
-logging.file = logs/api/rent-api.log
-
 # SPRING CLOUD
 spring.application.name = Rent-API
 
+<<<<<<< HEAD
```
Comment (beeduck) [inline Rent/Rent-API/src/main/resources/application.properties]: Fix this - merge conflicts

File: Rent/Rent-Data/src/main/java/com/rent/data/dataaccess/api/dao/listing/ListingImageDAO.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
+package com.rent.data.dataaccess.api.dao.listing;
```
Comment (beeduck) [inline Rent/Rent-Data/src/main/java/com/rent/data/dataaccess/api/dao/listing/ListingImageDAO.java]: Remove this file, Rent-Data should be dead

File: Rent/Rent-Data/src/main/java/com/rent/data/dataaccess/api/dao/listing/ListingImageDAOImpl.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
+package com.rent.data.dataaccess.api.dao.listing;
```
Comment (beeduck) [inline Rent/Rent-Data/src/main/java/com/rent/data/dataaccess/api/dao/listing/ListingImageDAOImpl.java]: Remove this file, Rent-Data should be dead

File: Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java (hunk 6 above)
Comment (AsadA128) [reply Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java]: /uuid already exists for get

File: Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java (hunk 7 above)
Comment (AsadA128) [reply Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java:47]: this also exists for get

File: Rent/Rent-API/src/main/java/com/rent/api/services/listing/ListingImageServiceImpl.java (hunk 8 above)
Comment (AsadA128) [reply Rent/Rent-API/src/main/java/com/rent/api/services/listing/ListingImageServiceImpl.java]: not this, the fileuploadcontroller

File: Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java (hunk 6 above)
Comment (beeduck) [reply Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java]: For GET, not DELETE

File: Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java (hunk 7 above)
Comment (beeduck) [reply Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java:47]: Remove the controller is what I mean, isn't it not necessary?

Comment (AsadA128) [reply Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java:47]: I think this needs to stay

Comment (AsadA128) [reply Rent/Rent-API/src/main/java/com/rent/api/controllers/listing/ListingImageController.java:47]: You need to delete by both image and listing, if someone wants to remove just an image, then they can delete it, if someone wants to delete an entire listing, then you need a way to delete all of them at once.

**Review by beeduck — APPROVED** [review]
+1

**Pull request merged**